polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -5080,33 +5073,34 @@ mlir::FunctionOpInterface MLIRASTConsumer::GetOrCreateMLIRFunction(
   }
   mlir::OpBuilder builder(module->getContext());
 
-  const auto createFunctionDecl = [&]() -> FunctionOpInterface {
-    const auto insert = [&](auto Function, auto Module, auto &Map) {
-      Module.push_back(Function);
-      Map[name] = Function;
-      return Function;
+  // Inject a function declaration in the deviceModule (a GPUModuleOp) if the
+  // function is a device function and in the module (a ModuleOp) otherwise.
+  const auto createFunction = [&](FunctionToEmit &F) -> FunctionOpInterface {
+    const auto insert = [name](auto function, auto module, auto &map) {
+      module.push_back(function);
+      map[name] = function;
+      return function;
     };
     auto funcType = builder.getFunctionType(types, rettypes);
-    const auto loc = getMLIRLocation(FD->getLocation());
+    Location loc = getMLIRLocation(F.getDecl().getLocation());
 
-    if (FD->hasAttr<SYCLKernelAttr>()) {
+    if (F.getDecl().hasAttr<SYCLKernelAttr>()) {
       auto function = builder.create<gpu::GPUFuncOp>(loc, name, funcType,
                                                      TypeRange{}, TypeRange{});
-      // SYCL kernels must be always located in a SYCL device context.
-      Context = FunctionContext::SYCLDevice;
-
       return insert(function, deviceModule, deviceFunctions);
     }
+
     auto function = builder.create<func::FuncOp>(loc, name, funcType);
-    switch (Context) {
+    switch (F.getContext()) {
     case FunctionContext::Host:
       return insert(function, *module, functions);
     case FunctionContext::SYCLDevice:
```

Comment:
@victor-eds Is it intentional to create `func.func` for SYCL device functions now places in `gpu.module`?
What are the differences between `func.func` and `gpu.func`?

Reply:
Should we change the function to `gpu.func` when it is placed in the SYCLDevice context?

Reply:
From the `gpu.func` documentation:
> The body of a GPU function, when launched, is executed by multiple work items. 
We cannot call `gpu.func`, just launching them. The `func.func`s we're introducing in the `gpu.module` are indeed regular functions we want to call from kernels, not launch them.

Reply:
ok, so SYCL device functions should remain as `func.func` in a `gpu.module` from what you are saying here.